File: Day6/Python/Main.py
import os

def main():
    input = os.path.dirname(os.getcwd())
    input = os.path.join(input, "input")
    if not os.path.exists(input):
        print(input + " does not exist!")
        input = os.path.join(os.path.dirname(input), "input.txt")
        print(f"trying {input} instead.")
        if not os.path.exists(input):
            print(input + " does not exist!")
            return;

    inputFile = open(input, 'r')
    globalGroupFound = 0
    globalFound = 0
    groupFoundEvery = []
    groupFoundAny = []
    found = []
    newline = False
    first = True
    for b in inputFile.read():
        if b == '\n':
            if newline:
                globalFound += len(groupFoundAny)
                globalGroupFound += len(groupFoundEvery)
                first = True
            else:
                if first:
                    groupFoundAny = found.copy()
                    groupFoundEvery = found.copy()
                    first = False
                else:
                    for a in found:
                        if not a in groupFoundAny:
                            groupFoundAny.append(a)
                    for a in groupFoundAny:
                        if not a in found and a in groupFoundEvery:
                            groupFoundEvery.remove(a)

                found.clear()
                newline = True
        else:
            newline = False
            if not b in found:
                found.append(b)

    globalFound += len(groupFoundAny)
    globalGroupFound += len(groupFoundEvery)

    print(f"The sum of all group part 1 answers is {globalFound}.")
    print(f"The sum of all group part 2 answers is {globalGroupFound}.")

if __name__ == '__main__':
    main()

File: Day6/Java/Main.java
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.HashSet;
import java.util.Set;

public class Main {

	public static void main(String[] args) throws IOException {
		File input = new File("input").getAbsoluteFile();
		if (!input.exists()) {
			System.out.println(input.getPath() + " does not exist!");
			System.out.println("trying " + input.getPath() + ".txt instead.");
			input = new File(input.getPath() + ".txt");
			if (!input.exists()) {
				System.err.println("neither " + input.getPath().substring(0, input.getPath().length() - 4) + ",");
				System.err.println("nor " + input.getPath() + " exists!");
				return;
			}
		}

		int globalGroupCount = 0;
		int globalCount = 0;
		Set<Byte> groupFoundAny = new HashSet<>();
		Set<Byte> groupFoundEvery = new HashSet<>();
		Set<Byte> found = new HashSet<>();
		boolean newline = false;
		boolean first = true;
		for (byte b : Files.readAllBytes(input.toPath())) {
			if (b == '\n') {
				if (newline) {
					globalCount += groupFoundAny.size();
					globalGroupCount += groupFoundEvery.size();
					groupFoundAny.clear();
					groupFoundEvery.clear();
					first = true;
				} else {
					groupFoundAny.addAll(found);
					if (first) {
						groupFoundEvery.addAll(found);
						first = false;
					} else {
						for (byte a : groupFoundAny) {
							if (!found.contains(a))
								if(groupFoundEvery.contains(a))
									System.out.println((char)a);
								groupFoundEvery.remove(a);
						}
					}
					found.clear();
					newline = true;
				}
			} else if (b == '\r') {
			} else {
				newline = false;
				found.add(b);
			}
		}

		globalCount += groupFoundAny.size();
		globalGroupCount += groupFoundEvery.size();

		System.out.printf("The sum of all group part 1 answers is %d.%n", globalCount);
		System.out.printf("The sum of all group part 2 answers is %d.%n", globalGroupCount);
	}
}
